lib/tesla/adapter/gun.ex

```diff
@@ -420,5 +429,17 @@ if Code.ensure_loaded?(:gun) do
         {:error, :body_too_large}
       end
     end
+
+    defp format_host(string) do
+      charlist = to_charlist(string)
+
+      case :inet.parse_address(charlist) do
+        {:error, :einval} ->
+          {:domain, :idna.encode(charlist)}
+
+        {:ok, ip} when is_tuple(ip) and tuple_size(ip) in [4, 8] ->
+          {:ip, ip}
+      end
+    end
```

Comment:
Would it make sense to have `domain_or_ss_a_string(uri)` instead? (name TBD) 

Seems like all use cases are to get encoded domain OR fallback to `uri.host` 🤔

Reply:
I renamed `format_host` function to return domain or ip. 
Created second function to return domain or fallback for ips. As `gun` wants domain or ip tuple to connect, but in other places needed fallback for ips.